include/revng/TypeNames/DependencyGraph.h

```diff
@@ -14,30 +16,96 @@
 struct TypeNode {
 
   /// A pointer to the associated model::TypeDefinition
-  const model::TypeDefinition *T;
+  const model::UpcastableType T;
 
   /// For each model::TypeDefinition we'll have nodes representing the type name
   /// or the full type, depending on this enum.
   enum Kind {
+    /// The TypeNode represents a declaration of a type name
     Declaration,
-    Definition
+    /// The TypeNode represents the full definition of a type name
+    Definition,
+    /// The TypeNode represents the forward declaration of an artificial struct
+    /// wrapper for the type referred to by T
+    ArtificialWrapperDeclaration,
+    /// The TypeNode represents the full definition of an artificial struct
+    /// wrapper for the type referred to by T
+    ArtificialWrapperDefinition,
   } K;
+
+  bool isDeclaration() const {
+    return K == Declaration or K == ArtificialWrapperDeclaration;
+  }
+
+  bool isDefinition() const {
+    return K == Definition or K == ArtificialWrapperDefinition;
+  }
+
+  bool isArtificial() const {
+    return K == ArtificialWrapperDefinition
+           or K == ArtificialWrapperDeclaration;
+  }
 };
 
 using TypeDependencyNode = BidirectionalNode<TypeNode>;
-using TypeKindPair = std::pair<const model::TypeDefinition *, TypeNode::Kind>;
-using TypeToDependencyNodeMap = std::map<TypeKindPair, TypeDependencyNode *>;
-using TypeVector = TrackingSortedVector<model::UpcastableTypeDefinition>;
 
 /// Represents the graph of dependencies among types
-struct DependencyGraph : public GenericGraph<TypeDependencyNode> {
+class DependencyGraph : public GenericGraph<TypeDependencyNode> {
+private:
+  /// A pair of associated nodes that are respectively the declaration and the
+  /// definition of the same type.
+  struct AssociatedNodes {
+    TypeDependencyNode *Declaration;
+    TypeDependencyNode *Definition;
+  };
+
+  /// A map type that maps a model::TypeDefinition to a pair of
+  /// TypeDependencyNodes, representing respectively the declaration and the
+  /// definition of such TypeDefinition.
+  using DefinitionToNodesMap = std::unordered_map<const model::TypeDefinition *,
+                                                  AssociatedNodes>;
 
-  void addNode(const model::TypeDefinition *T);
+  /// A map type that maps a model::UpcastableType to a pair of
+  /// TypeDependencyNodes, representing respectively the declaration and the
+  /// definition of an artificial struct wrapper for that UpcastableType.
+  using WrappedTypeToNodesMap = std::map<const model::UpcastableType,
+                                         AssociatedNodes>;
 
-  const TypeToDependencyNodeMap &TypeNodes() const { return TypeToNode; }
+public:
+  /// A internal factory class used to build a DependencyGraph
+  class Builder;
 
 private:
-  TypeToDependencyNodeMap TypeToNode;
+  /// Maps a model::TypeDefinition * to an AssociatedNodes, representing
```

Comment:
`an AssociatedNodes` sounds weird, I'd go with just `AssociatedNodes` or maybe even `to its AssociatedNodes` (same below).